crates/rpc/src/v04/method/trace_block_transactions.rs

```diff
@@ -0,0 +1,206 @@
+    fn from(value: CallError) -> Self {
+        match value {
+            CallError::ContractNotFound | CallError::InvalidMessageSelector => Self::Internal(
+                anyhow::anyhow!("Failed to trace the transaction: {value:?}"),
+            ),
+            CallError::Reverted(e) => Self::Internal(anyhow::anyhow!("Transaction reverted: {e}")),
+            CallError::Internal(e) => Self::Internal(e),
+        }
+    }
+}
+
+impl From<JoinError> for TraceBlockTransactionsError {
+    fn from(e: JoinError) -> Self {
+        Self::Internal(anyhow::anyhow!("Join error: {e}"))
+    }
+}
+
+pub async fn trace_block_transactions(
+    context: RpcContext,
+    input: TraceBlockTransactionsInput,
+) -> Result<TraceBlockTransactionsOutput, TraceBlockTransactionsError> {
+    let transactions: Vec<_> = {
+        let mut db = context.storage.connection()?;
+        tokio::task::spawn_blocking(move || {
+            let tx = db.transaction()?;
+
+            let (transactions, _): (Vec<_>, Vec<_>) = tx
+                .transaction_data_for_block(BlockId::Hash(input.block_hash))?
+                .ok_or(TraceBlockTransactionsError::InvalidBlockHash)?
+                .into_iter()
+                .unzip();
+
+            let hashes = transactions.iter().map(|tx| tx.hash()).collect::<Vec<_>>();
+
+            let transactions = transactions
+                .into_iter()
+                .map(|transaction| compose_executor_transaction(transaction, &tx))
+                .collect::<anyhow::Result<Vec<_>, _>>()?;
+
+            Ok::<_, TraceBlockTransactionsError>(
+                hashes.into_iter().zip(transactions.into_iter()).collect(),
+            )
+        })
+        .await??
+    };
+
+    let execution_state =
+        crate::executor::execution_state(context, pathfinder_common::BlockId::Latest, None).await?;
+
+    let traces = tokio::task::spawn_blocking(move || {
+        pathfinder_executor::trace_all(execution_state, transactions)
+    })
+    .await??;
+
+    let result = traces
+        .into_iter()
+        .map(|(hash, trace)| Trace {
+            transaction_hash: hash,
+            trace_root: trace.into(),
+        })
+        .collect();
+
+    Ok(TraceBlockTransactionsOutput(result))
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use crate::{
+        v02::method::call::FunctionCall,
+        v04::method::simulate_transactions::dto::{
+            CallType, DeployAccountTxnTrace, EntryPointType, FunctionInvocation,
+        },
+    };
+
+    use super::*;
+
+    use pathfinder_common::{
+        class_hash, BlockHeader, BlockNumber, BlockTimestamp, CallParam, ContractAddress,
+        EntryPoint,
+    };
+    use pathfinder_storage::Storage;
+    use starknet_gateway_types::reply as gateway;
+
+    pub(crate) fn setup_trace_test() -> anyhow::Result<(Storage, gateway::Block, TransactionTrace)>
+    {
+        const TEST_BLOCK: &str = include_str!("../../../fixtures/trace/block.json");
```

Comment:
We already have a fairly extensive test setup in `pathfinder_rpc::test_setup` and `simulate_transactions`. Can't we simply re-use those in these trace tests somehow? We already have expected traces, etc for more transactions types so I though re-using those would be simple?